### PR TITLE
Remove unused `base64.encode` overload

### PR DESCRIPTION
**Pull request by AmjadHD**

### CODE REVIEW

File: lib/pure/base64.nim [hunk 1]
```diff
@@ -144,41 +144,38 @@ template encodeImpl() {.dirty.} =
...
+#   ## which substitutes `-` instead of `+` and `_` instead of `/`.
+#   ## * https://en.wikipedia.org/wiki/Base64#URL_applications
+#   ## * https://tools.ietf.org/html/rfc4648#page-7
+#   ##
+#   ## **See also:**
+#   ## * `encode proc<#encode,openArray[T]>`_ for encoding an openarray
+#   ## * `decode proc<#decode,string>`_ for decoding a string
+#   runnableExamples:
+#     assert encode("Hello World") == "SGVsbG8gV29ybGQ="
+#   encodeImpl()
```
Comment (juancarlospaco) [inline lib/pure/base64.nim]: Do not comment out whole procs, remove it or improve it, but not let it commented out like that IMHO. :thinking:

Comment (AmjadHD) [reply lib/pure/base64.nim]: Oh, I thought I removed it.

**Comment by Varriount** [comment]
Do we know why this procedure was added in the first place?

**Comment by AmjadHD** [comment]
There was the string overload, then the openarray overload was added in 6901a8cb13d19a3d251911af240c5f2a54809eba (https://github.com/nim-lang/Nim/pull/646), Ig the author wasn't aware that `string` is compatible with `openArray[char]`.

**Comment by AmjadHD** [comment]
Is it ok to have an overload for integer types of arbitrary sizes. Shouldn't the input be byte strings.

**Comment by Varriount** [comment]
Please add a changelog entry, then this can be merged.

> Is it ok to have an overload for integer types of arbitrary sizes. Shouldn't the input be byte strings.

Is this a question? And if so, what code are you referencing?

**Comment by AmjadHD** [comment]
> Please add a changelog entry, then this can be merged.

Is this needed, from the user's perspective nothing changed.
> > Is it ok to have an overload for integer types of arbitrary sizes. Shouldn't the input be byte strings.
> 
> Is this a question? And if so, what code are you referencing?

Yes, https://github.com/nim-lang/Nim/pull/20369/files#diff-cc4f63a53ac195237326427c11b84c89967c417d474b4cd55ad54b48a09c9a1cL144.
This overload takes arbitrary integer types and wraps the values that do not fit in 1 byte, Is this behaviour desired or should it error instead ?

**Comment by Varriount** [comment]
Can you give an example of what happens with 1-byte integers, vs multi-byte integers?

**Comment by AmjadHD** [comment]
```nim
import std/base64

var x = @[72, 101, 108, 108, 111] # Hello
let e = encode(x)
x[0] += 256
assert encode(x) == e
```

**Comment by Varriount** [comment]
Hm. I don't know.
I would expect is one of the following:
  - When given a multi-byte integer, `encode` would output multiple characters.
  - `encode` would not accept multi-byte integers.

Both are breaking changes, though the current behavior is arguably a bug. I would implement #2 first, and see how many, if any, packages that breaks.

**Comment by AmjadHD** [comment]
I tried `T: range[0..255]|uint8|char` but this doesn't work even for `encode [10]`.

**Comment by arnetheduck** [comment]
for reference, the API we use is: https://github.com/status-im/nim-stew/blob/master/stew/base64.nim (notably, there are many "base64" alphabet standards)

**Comment by AmjadHD** [comment]
Should I change `T: byte|char` then and deprecate the `T: SomeInteger` overload ?

**Comment by AmjadHD** [comment]
> I tried `T: range[0..255]|uint8|char` but this doesn't work even for `encode [10]`.

Why doesn't it work ?

**Comment by AmjadHD** [comment]
cc @Varriount ^.

**Comment by Araq** [comment]
Instead of `T: range[0..255]|uint8|char` use `T: byte|char`

**Comment by AmjadHD** [comment]
@Araq, What about code like `encode @[111, 113, 32]` ? should it be deprecated ?

**Comment by AmjadHD** [comment]
@Araq is this fine ?

**Comment by Araq** [comment]
> @Araq, What about code like encode @[111, 113, 32] ? should it be deprecated ?

I think so but there is no RFC for it so better leave it as it is.

**Comment by AmjadHD** [comment]
Can this be merged now ?

File: lib/pure/base64.nim [hunk 2]
```diff
@@ -144,40 +146,23 @@ template encodeImpl() {.dirty.} =
 proc encode*[T: SomeInteger|char](s: openArray[T], safe = false): string =
```
Comment (Araq) [inline lib/pure/base64.nim]: So ... you left this one as `SomeInteger`? Bad idea.

Comment (AmjadHD) [reply lib/pure/base64.nim]: > > @Araq, What about code like encode @[111, 113, 32] ? should it be deprecated ?
> 
> I think so but there is no RFC for it so better leave it as it is.

So code like that just breaks ?

Comment (Varriount) [reply lib/pure/base64.nim]: Wait, so can you explain why this procedure was left as-is?

Comment (AmjadHD) [reply lib/pure/base64.nim]: > Wait, so can you explain why this procedure was left as-is?

> > @Araq, What about code like encode @[111, 113, 32] ? should it be deprecated ?
> 
> 
> I think so but there is no RFC for it so better leave it as it is.
> 

" leave it as it is"

Comment (AmjadHD) [reply lib/pure/base64.nim]: @Araq please clarify what you want:
1. `T: byte|char`
2. `T: byte|char` and deprecate `T: SomeInteger and not byte`
3. `T: SomeInteger|char`

Comment (Araq) [reply lib/pure/base64.nim]: (2)

**Comment by AmjadHD** [comment]
@Araq merge this please.

File: lib/pure/base64.nim [hunk 3]
```diff
@@ -141,43 +143,26 @@ template encodeImpl() {.dirty.} =
...
-  ## * https://en.wikipedia.org/wiki/Base64#URL_applications
-  ## * https://tools.ietf.org/html/rfc4648#page-7
-  ##
-  ## **See also:**
-  ## * `encode proc<#encode,openArray[T]>`_ for encoding an openarray
-  ## * `decode proc<#decode,string>`_ for decoding a string
-  runnableExamples:
-    assert encode("Hello World") == "SGVsbG8gV29ybGQ="
+proc encode*[T: SomeInteger and not byte](s: openArray[T], safe = false): string  =
+  {.warning: "Deprecated Element type of `s`. use `byte` or `char` instead".}
```
Comment (Araq) [inline lib/pure/base64.nim]:
```suggestion
proc encode*[T: SomeInteger and not byte](s: openArray[T], safe = false): string {.deprecated: "use `byte` or `char` instead".} =
```

**Pull request merged**